core/src/query.rs

```diff
@@ -1,77 +1,551 @@
+                    .into_iter()
+                    .map(|t| {
+                        match t {
+                            // If one of the BSON types is Array, move the items
+                            // information into the schema that will be nested in
+                            // the top-level any_of and set the top-level items to
+                            // None.
+                            BsonTypeName::Array => {
+                                let nested_items = items.clone();
+                                items = None;
+                                SimplifiedJsonSchema {
+                                    bson_type: Some(t),
+                                    items: nested_items,
+                                    ..Default::default()
+                                }
+                            }
+                            // If one of the BSON types is Object, move the
+                            // properties, required, and additional_properties
+                            // information into the schema that will be nested
+                            // in the top-level any_of and set the top-level
+                            // fields to None.
+                            BsonTypeName::Object => {
+                                let nested_properties = properties.clone();
+                                let nested_required = required.clone();
+                                let nested_additional_properties = additional_properties;
+                                properties = None;
+                                required = None;
+                                additional_properties = false;
+                                SimplifiedJsonSchema {
+                                    bson_type: Some(t),
+                                    properties: nested_properties,
+                                    required: nested_required,
+                                    additional_properties: nested_additional_properties,
+                                    ..Default::default()
+                                }
+                            }
+                            // If we encounter any other BSON type, simply create
+                            // a SimplifiedJsonSchema with the relevant bson_type.
+                            t => SimplifiedJsonSchema {
+                                bson_type: Some(t),
+                                ..Default::default()
+                            },
+                        }
+                    })
+                    .collect::<Vec<SimplifiedJsonSchema>>();
+
+                // Extend the top-level any_of with the new, additional
+                // schemas created here from the Multiple BSON types.
+                any_of = match any_of {
+                    None => Some(additional_any_of_schemas),
+                    Some(mut v) => {
+                        v.append(&mut additional_any_of_schemas);
```

Comment:
Note that I am not removing duplicate types from the any_of. I could do that for sure, but duplicates don't actually impact correctness the same way nested any_ofs do. As in, there are other parts of this code that rely on `SimplifiedJsonSchema` conforming to the invariant of not having nested any_ofs, but there is nothing that relies on any_ofs not containing duplicates.